packages/ai-ide/src/browser/task-context-file-storage-service.ts

```diff
@@ -0,0 +1,174 @@
+import { PreferenceService } from '@theia/core/lib/browser';
+import { FileService } from '@theia/filesystem/lib/browser/file-service';
+import { WorkspaceService } from '@theia/workspace/lib/browser';
+import * as yaml from 'js-yaml';
+import { FileChange, FileChangeType } from '@theia/filesystem/lib/common/files';
+import { TASK_CONTEXT_STORAGE_DIRECTORY_PREF } from './workspace-preferences';
+import { BinaryBuffer } from '@theia/core/lib/common/buffer';
+
+@injectable()
+export class TaskContextFileStorageService implements TaskContextStorageService {
+    @inject(InMemoryTaskContextStorage) protected readonly inMemoryStorage: InMemoryTaskContextStorage;
+    @inject(PreferenceService) protected readonly preferenceService: PreferenceService;
+    @inject(WorkspaceService) protected readonly workspaceService: WorkspaceService;
+    @inject(FileService) protected readonly fileService: FileService;
+    protected readonly onDidChangeEmitter = new Emitter<void>();
+    readonly onDidChange = this.onDidChangeEmitter.event;
+
+    protected getStorageLocation(): URI | undefined {
+        if (!this.workspaceService.opened) { return; }
+        const values = this.preferenceService.inspect(TASK_CONTEXT_STORAGE_DIRECTORY_PREF);
+        const configuredPath = values?.globalValue === undefined ? values?.defaultValue : values?.globalValue;
+        if (!configuredPath || typeof configuredPath !== 'string') { return; }
+        const asPath = new Path(configuredPath);
+        return asPath.isAbsolute ? new URI(configuredPath) : this.workspaceService.tryGetRoots().at(0)?.resource.resolve(configuredPath);
+    }
+
+    @postConstruct()
+    protected init(): void {
+        this.watchStorage();
+        this.preferenceService.onPreferenceChanged(e => {
+            if (e.affects(TASK_CONTEXT_STORAGE_DIRECTORY_PREF)) { this.watchStorage(); }
+        });
+    }
+
+    protected toDisposeOnStorageChange?: DisposableCollection;
+    protected async watchStorage(): Promise<void> {
+        this.toDisposeOnStorageChange?.dispose();
+        this.toDisposeOnStorageChange = undefined;
+        const newStorage = this.getStorageLocation();
+        if (!newStorage) { return; }
+        this.toDisposeOnStorageChange = new DisposableCollection(
+            this.fileService.watch(newStorage),
+            this.fileService.onDidFilesChange(event => {
+                const relevantChanges = event.changes.filter(candidate => newStorage.isEqualOrParent(candidate.resource));
+                this.handleChanges(relevantChanges);
+            }),
+            { dispose: () => this.clearInMemoryStorage() },
+        );
+        await this.cacheNewTasks(newStorage);
+    }
+
+    protected async handleChanges(changes: FileChange[]): Promise<void> {
+        await Promise.all(changes.map(change => {
+            switch (change.type) {
+                case FileChangeType.DELETED: return this.deleteFileReference(change.resource);
+                case FileChangeType.ADDED:
```

Comment:
Can we implement this case? If the user manually adds a file to the directory, we should consider it.

Reply:
It falls through and is handled the same as modification: the file is read and the summary added / updated.